packages/block-library/src/table/edit.js

```diff
@@ -0,0 +1,294 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { Fragment, Component } from '@wordpress/element';
+import { InspectorControls, BlockControls, RichText } from '@wordpress/editor';
+import { __ } from '@wordpress/i18n';
+import {
+	PanelBody,
+	ToggleControl,
+	TextControl,
+	Button,
+	Toolbar,
+	DropdownMenu,
+} from '@wordpress/components';
+
+import {
+	createTable,
+	updateCellContent,
+	insertRow,
+	deleteRow,
+	insertColumn,
+	deleteColumn,
+} from './state';
+
+const { parseInt } = window;
+
+export default class TableEdit extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.onCreateTable = this.onCreateTable.bind( this );
+		this.onChangeFixedLayout = this.onChangeFixedLayout.bind( this );
+		this.createOnChange = this.createOnChange.bind( this );
+		this.onChangeInitialColumnCount = this.onChangeInitialColumnCount.bind( this );
+		this.onChangeInitialRowCount = this.onChangeInitialRowCount.bind( this );
+		this.renderSection = this.renderSection.bind( this );
+		this.getTableControls = this.getTableControls.bind( this );
+		this.createOnDeleteRow = this.createOnDeleteRow.bind( this );
+		this.createOnDeleteColumn = this.createOnDeleteColumn.bind( this );
+		this.createOnInsertRow = this.createOnInsertRow.bind( this );
+		this.createOnInsertColumn = this.createOnInsertColumn.bind( this );
+		this.createOnFocus = this.createOnFocus.bind( this );
+
+		this.state = {
+			initialRowCount: 2,
+			initialColumnCount: 2,
+			selectedCell: null,
+		};
+	}
+
+	onChangeInitialColumnCount( initialColumnCount ) {
+		this.setState( { initialColumnCount } );
+	}
+
+	onChangeInitialRowCount( initialRowCount ) {
+		this.setState( { initialRowCount } );
+	}
+
+	onCreateTable() {
+		const { setAttributes } = this.props;
+		let { initialRowCount, initialColumnCount } = this.state;
+
+		initialRowCount = parseInt( initialRowCount, 10 ) || 2;
+		initialColumnCount = parseInt( initialColumnCount, 10 ) || 2;
+
+		setAttributes( createTable( {
+			rowCount: initialRowCount,
+			columnCount: initialColumnCount,
+		} ) );
+	}
+
+	onChangeFixedLayout() {
+		const { attributes, setAttributes } = this.props;
+		const { hasFixedLayout } = attributes;
+
+		setAttributes( { hasFixedLayout: ! hasFixedLayout } );
+	}
+
+	createOnChange( { section, rowIndex, columnIndex } ) {
+		return ( content ) => {
+			const { attributes, setAttributes } = this.props;
+			setAttributes( updateCellContent( attributes, {
+				section,
+				rowIndex,
+				columnIndex,
+				content,
+			} ) );
+		};
+	}
+
+	createOnInsertRow( { section, rowIndex }, delta ) {
+		return () => {
+			const { attributes, setAttributes } = this.props;
+			rowIndex += delta;
+			setAttributes( insertRow( attributes, { section, rowIndex } ) );
```

Comment:
`rowIndex` is bound to `createOnInsertRow`'s lexical scope, but incremented when the returned anonymous function is called. This might result in a bug where clicking _Insert Row_ inserts a new row two down from the current row because `rowIndex` was incremented twice.

I think that, in practice, this bug won't happen because inserting a row will cause React to call `render()` which in turn calls `createOnInsertRow()` and re-initialises `rowIndex`. 

Still, I think it's safer to not rely on this behaviour and instead avoid the mutating `+=` operator here.

```js
const { attributes, setAttributes } = this.props;
setAttributes( insertRow( attributes, { section, rowIndex: rowIndex + delta } ) );
```